velox/dwio/common/DirectDecoder.h

```diff
@@ -93,7 +93,17 @@ class DirectDecoder : public IntDecoder<isSigned> {
       } else if constexpr (std::is_same_v<
                                typename Visitor::DataType,
                                int128_t>) {
-        toSkip = visitor.process(super::template readInt<int128_t>(), atEnd);
+        if (super::numBytes_ != 12) {
+          toSkip = visitor.process(super::template readInt<int128_t>(), atEnd);
+        } else {
+          // Reads INT96 timestamp as int128_t type and extracts the days and
+          // nanos.
+          const int128_t encoded = super::template readInt<int128_t>();
+          const int32_t days = encoded & ((1ULL << 32) - 1);
+          const uint64_t nanos = static_cast<uint64_t>(encoded >> 32);
+          auto ts = Timestamp::fromDaysAndNanos(days, nanos);
```

Comment:
Integer decoder should not hand out timestamps.  Can we just read int128 as it is, and doing this transformation in `TimestampColumnReader::read`?

Reply:
This value will be used in 'velox::common::applyFilter', and the raw Int96 needs to be converted as Timestamp for comparison. Changed to return wrapped int128_t in the 'readInt96' function, and this special handling could be removed. Do you think it makes sense? Thanks.

https://github.com/facebookincubator/velox/pull/10850/files#diff-88067c13057d85cf40dc2ca5dd620c342946f606f9f0976696ebc88a3f1f0713R475-R476